pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,309 @@
+        gradient and grow the circuit.
+
+        Args:
+            circuit (Any): the user-defined hyperparameter :math:`\eta`
+            stepsize (float): the user-defined hyperparameter :math:`\eta`
+            restriction (qml.Hamiltonian): Restrict the Lie algebra to a corresponding subspace of
+            the full Lie algebra. This restriction should be passed in the form of a
+            `qml.Hamiltonian`.
+            exact (bool): Flag that indicates wether we approximate the Lie gradient with a
+            Trotterization or calculate the exact evolution via a matrix exponential. The latter is
+            not quantum friendly and can only be done in simulation.
+
+            **kwargs:
+
+        **Examples:**
+
+        Define a Hamiltonian cost function to minimize.
+        >>> hamiltonian = qml.Hamiltonian(coeffs=[-1.]*3,
+        ...observables=[qml.PauliX(0), qml.PauliZ(1), qml.PauliY(0)@qml.PauliX(1)])
+        Create an initial state and return the expectation value of the Hamiltonian.
+        >>> @qml.qnode(qml.device("default.qubit", wires=2))
+        ... def quant_fun():
+        ...     qml.RX(0.1, wires=[0])
+        ...     qml.RY(0.5, wires=[1])
+        ...     qml.CNOT(wires=[0,1])
+        ...     qml.RY(0.6, wires=[0])
+        ...     return qml.expval(hamiltonian)
+
+        Instatiate the optimizer with the initial circuit and the cost function. Set the stepsize
+        accordingly.
+        >>> opt = LieGradientOptimizer(circuit = quant_fun, stepsize=0.1)
+        Applying 10 steps gets us close the ground state of E=-2.23
+        >>> for step in range(10):
+        ...    print(step)
+        ...    cost = opt.step_and_cost()
+        ...    print(cost)
+
+        """
+        if not isinstance(circuit, qml.QNode):
+            raise TypeError(
+                f"`circuit` must be a `qml.QNode`, " f"received {type(circuit)} "
+            )
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        if not isinstance(circuit.func().obs, qml.Hamiltonian):
+            raise TypeError(
+                f"`circuit` must return the expectation value of a `qml.Hamiltonian`,"
+                f" "
+                f"received {type(circuit.func().obs)} "
+            )
+        self.nqubits = max(circuit.device.wires) + 1
+
+        if self.nqubits > 4:
+            print(
+                "WARNING: The exact Lie gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow."
+            )
+        if restriction is not None and not isinstance(restriction, qml.Hamiltonian):
+            raise TypeError(f"`restriction` must be a `qml.Hamiltonian`, received {type(restriction)}")
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators(restriction)
+        self.exact = exact
+        self.hamiltonian = circuit.func().obs
```

Comment:
if you move this line to above line 180, you won't need to evaluate the circuit function twice 🙂